src/catalog/sql.rs

```diff
@@ -14,54 +14,89 @@ use rusqlite::types::{FromSql, FromSqlError, ToSql, ToSqlOutput, Value, ValueRef
 use serde::{Deserialize, Serialize};
 
 use expr::GlobalId;
+use ore::cast::CastFrom;
 
 use crate::error::{Error, ErrorKind};
 use crate::names::{DatabaseSpecifier, FullName};
 
 const APPLICATION_ID: i32 = 0x1854_47dc;
 
-const SCHEMA: &str = "
-CREATE TABLE gid_alloc (
-    next_gid integer NOT NULL
-);
-
-CREATE TABLE databases (
-    id   integer PRIMARY KEY,
-    name text NOT NULL UNIQUE
-);
-
-CREATE TABLE schemas (
-    id          integer PRIMARY KEY,
-    database_id integer REFERENCES databases,
-    name        text NOT NULL,
-    UNIQUE (database_id, name)
-);
-
-CREATE TABLE items (
-    gid        blob PRIMARY KEY,
-    schema_id  integer REFERENCES schemas,
-    name       text NOT NULL,
-    definition blob NOT NULL,
-    UNIQUE (schema_id, name)
-);
-
-CREATE TABLE timestamps (
-    sid blob NOT NULL,
-    vid blob NOT NULL,
-    pcount blob NOT NULL,
-    pid blob NOT NULL,
-    timestamp integer NOT NULL,
-    offset blob NOT NULL,
-    PRIMARY KEY (sid, vid, pid, timestamp)
-);
-
-INSERT INTO gid_alloc VALUES (1);
-INSERT INTO databases VALUES (1, 'materialize');
-INSERT INTO schemas VALUES
-    (1, NULL, 'mz_catalog'),
-    (2, NULL, 'pg_catalog'),
-    (3, 1, 'public');
-";
+/// Schema migrations for the on-disk state.
+const MIGRATIONS: &[&str] = &[
+    // Creates initial schema.
+    //
+    // Introduced for v0.1.0.
+    "CREATE TABLE gid_alloc (
+         next_gid integer NOT NULL
+     );
+
+     CREATE TABLE databases (
+         id   integer PRIMARY KEY,
+         name text NOT NULL UNIQUE
+     );
+
+     CREATE TABLE schemas (
+         id          integer PRIMARY KEY,
+         database_id integer REFERENCES databases,
+         name        text NOT NULL,
+         UNIQUE (database_id, name)
+     );
+
+     CREATE TABLE items (
+         gid        blob PRIMARY KEY,
+         schema_id  integer REFERENCES schemas,
+         name       text NOT NULL,
+         definition blob NOT NULL,
+         UNIQUE (schema_id, name)
+     );
+
+     CREATE TABLE timestamps (
+         sid blob NOT NULL,
+         vid blob NOT NULL,
+         timestamp integer NOT NULL,
+         offset blob NOT NULL,
+         PRIMARY KEY (sid, vid, timestamp)
+     );
+
+     INSERT INTO gid_alloc VALUES (1);
+     INSERT INTO databases VALUES (1, 'materialize');
+     INSERT INTO schemas VALUES
+         (1, NULL, 'mz_catalog'),
+         (2, NULL, 'pg_catalog'),
+         (3, 1, 'public');",
+    // Adjusts timestamp table to support multi-partition Kafka topics.
+    //
+    // Introduced for v0.1.4.
+    //
+    // ATTENTION: this migration blows away data and must not be used as a model
+    // for future migrations! It is only acceptable now because we have not yet
+    // made any consistency promises to users.
+    "DROP TABLE timestamps;
+     CREATE TABLE timestamps (
+        sid blob NOT NULL,
+        vid blob NOT NULL,
+        pcount blob NOT NULL,
+        pid blob NOT NULL,
```

Comment:
I have no idea what I'm talking about, but is it not possible to alter this table with default values (maybe 1 and 0?) and then drop the default once they've been filled in?

The main thing that I have no idea about obviously being what valid temporary defaults would be.

Reply:
It might also be easier to simply stop writing timestamp info to the catalog (we use it to replay a stream consistently). Frank was reluctant to store that information in SQL Lite. As sink support increases, we could just write it to a sink?

Reply:
SQLite doesn't support changing a table's primary key, so we have no choice but to drop and recreate: https://www.sqlite.org/lang_altertable.html

We could try harder to preserve the existing data for sure, but didn't seem worth it.